drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,438 @@
+};
+
+static struct linear_range ldo3_ranges[] = {
+	LINEAR_RANGE_INIT(0u, 0u, 0u, 0u),
+	LINEAR_RANGE_INIT(AXP192_LDO3_VOLTAGE_MIN_UV, AXP192_LDO3_VOLTAGE_STEP_UV,
+			  AXP192_LDO3_VOLTAGE_RANGE_MIN, AXP192_LDO3_VOLTAGE_RANGE_MAX),
+};
+
+static const struct regulator_axp192_desc ldo3_desc = {
+	.enable_reg = AXP192_REG_DCDC123_LDO23_CONTROL,
+	.enable_mask = AXP192_LDO3_ENABLE_REG_MASK,
+	.enable_val = AXP192_LDO3_ENABLE_REG_VAL,
+	.vsel_reg = AXP192_REG_LDO23_VOLTAGE,
+	.vsel_mask = AXP192_LDO3_VOLTAGE_REG_MASK,
+	.vsel_bitpos = AXP192_LDO3_VOLTAGE_REG_BITPOS,
+	.max_ua = AXP192_LDO3_CURRENT_MAX_UA,
+	.ranges = ldo3_ranges,
+	.num_ranges = ARRAY_SIZE(ldo3_ranges),
+};
+
+int axp192_enable(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->mfd->config;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "Enabling regulator");
+	LOG_INST_DBG(config->log, "[0x%02x]=0x%02x mask=0x%02x", (uint32_t)config->desc->enable_reg,
+		     (uint32_t)config->desc->enable_val, (uint32_t)config->desc->enable_mask);
+
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, config->desc->enable_val);
+
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to enable regulator");
+	}
+
+	return ret;
+}
+
+int axp192_disable(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->mfd->config;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "Disabling regulator");
+	LOG_INST_DBG(config->log, "[0x%02x]=0 mask=0x%x", (uint32_t)config->desc->enable_reg,
+		     (uint32_t)config->desc->enable_mask);
+
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, 0u);
+
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to disable regulator");
+	}
+
+	return ret;
+}
+
+unsigned int axp192_count_voltages(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_values_count(config->desc->ranges, config->desc->num_ranges);
+}
+
+int axp192_list_voltage(const struct device *dev, unsigned int idx, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, idx,
+					    volt_uv);
+}
+
+int axp192_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->mfd->config;
+	uint16_t idx = 0;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "voltage = [min=%d, max=%d]", min_uv, max_uv);
+
+	/* configure voltage */
+	if (max_uv == 0) {
+		/* disable regulator if voltage is 0 */
+		ret = axp192_disable(dev);
+		if (ret != 0) {
+			return ret;
+		}
```

Comment:
this behavior is not documented in the interface, please drop

Reply:
I was following the implementation in pca9420 regulator driver.
What is the proper way to disable a regulator on startup?

Reply:
It does only support disabling the regulator.
And you are saying that there is no solution to disable a regulator via DT and I need to add a specific pre_app service for a board to disable a regulator?

Reply:
is regulator enabled by default? If so I guess you can add disable it during init unless regulator-boot-on/always-on is set (see regulator_fixed for example).

Reply:
Removed the option to disable regulator by allowing 0V. Imo the whole driver structure is lagging an option to explicitly disable a regulator on startup. With the boot-on and always-on option it is only possible to enable it.